### DISCUSSION TITLE
Best settings for The Legend of Zelda: Twilight Princess to fix bloom

### DISCUSSION THREAD

**Discussion by KoB-Kirito** [TryTwo/dolphin#8] [Q&A]
Can you recommend settings for Twilight Princess? The bloom in that game is especially broken.

**Comment by KoB-Kirito** [comment]
I guess I can provide my testings so far.. Will also shine some light on the problem.
I only tested the first shadow realm scene yet though (meeting Midna), so all might break later still.

How the scene looks on hardware, and should look I guess:
https://youtu.be/CB8ED-Ja1BU?t=67

How most people play it on dolphin: [6x internal (4k), scaled EFB on]
![6x 4k efb on](https://github.com/user-attachments/assets/e1d20859-e8c4-45be-999b-aabdf87ad4a5)

The EFB used for the bloom isn't scaled correctly resulting in weird ghosting.
You can kinda fix that on dolphin by playing..
..with native resolution: [1x internal]
![native efb off](https://github.com/user-attachments/assets/f20e7988-9c36-4195-b4ec-368fa9c8ef3f)

..or scaled efb off: [6x internal (4k), scaled EFB off]
![6x 4k efb off](https://github.com/user-attachments/assets/1eedf192-c698-4dd9-a69f-b72bf91fde06)

But that gives you a blurry image or causes other glitches. Also the intended ghosting (left right) is kinda too prominent at that resolution, imho.
So a lot just turn bloom off completely: [6x internal (4k), bloom disabled (AR code)]
![6x 4k bloom disabled ar code](https://github.com/user-attachments/assets/59b69428-c37f-412e-87ac-08a13d638800)

I also played like that for a while. It looks super crisp at first and I get if some prefer it that way. But all shadow areas look quite cold and dull like this. It also removes all glow from characters like Midna, Wolf Link and monsters. So I try to bring it back but good looking.
There are gecko codes on the wiki site to fix bloom. I could not test those btw, cause they don't work on the PAL version I play. Couldn't find working versions, tell me if you do. Maybe they also look good.
Until then this EFB slider tool seems to be very promising, awesome work!

Here are the results after some fiddling: [6x internal (4k), EFB on, downscale on: 160, edit on: 2, 100%]
![6x 4k efb on downscale 160 edit 2](https://github.com/user-attachments/assets/1ac9d1af-a779-4ee4-9bb6-f6fafbe7ad76)

[6x internal (4k), EFB on, downscale on: 160, edit on: 2, 80%]
![6x 4k efb on downscale 160 edit 2 80](https://github.com/user-attachments/assets/ff09530f-53a8-4212-83d1-a461d0e4675b)

I think I like the last one the most. That would be these settings:
![BloomControl2024_1 2](https://github.com/user-attachments/assets/191bad33-c882-4c97-8566-ed5d5e4f8a58)

**Comment by TryTwo** [comment]
Edit Bloom had the best results for me. It makes it higher res.  I would try with Downscale off, as it usually isn't needed if Edit Bloom is On. Your other settings seem fine. I think you've essentially answered your own question.

**Comment by KoB-Kirito** [comment]
Kinda, but I'm still not fully satisfied :D Maybe someone finds better settings, so I hope. At least this discussion could be helpful to others in search for an solution. I couldn't find much. Only found your tool here after days of searching.

If I disable downscale, the ghosting gets quite prominent again. I tried it without though:
This is with edit enabled only. Width 160, radius 4, strength 60:
![6x 4k bloom 160 4 60](https://github.com/user-attachments/assets/bb4b61a1-dfb3-492e-94b3-784e7ac413c9)
 Not sure I like that more, as the pulses blur the image more. (hence why I take gifs instead of still images to make that effect more visible).
 
 ALSO, I fiddled with the Wii U version. There was no bloom setting there, but it renders correctly at least, just very strong.
 But now I found a mod for bloom, it was just a bit hidden for some reason. It lets you alter the strength and also the color, which looks quite cool. I took screenshots of that here:
 https://github.com/cemu-project/cemu_graphic_packs/issues/508#issuecomment-2348661868
 ~~There you can also get the graphic pack for cemu.~~
EDIT: For someone finding this later: Use this new pack instead: https://github.com/cemu-project/cemu_graphic_packs/pull/642
 
 The HD remaster has it's own flaws and benefits, but in regard of bloom it might be the better option currently.

**Comment by TryTwo** [comment]
Hmm thank you.  I typically test things at lower IRs, but I see how downscaling bloom at higher IRs helps now.  I wonder if there should be a separate IR resolution for bloom... not sure what it'd do.

It's also possible I could use better math for the radius.  I just do basic averaging, but gaussian could be a good improvement at a high radius.
